File: ros2_ws/src/blindspot_arm/blindspot_arm/kin.py
"""Forward kinematics and geometric Jacobian for a serial chain, from URDF.

Pure numpy, no KDL. Written so it can be checked against TF rather than
trusted: verify_against_tf() compares this FK to what robot_state_publisher
reports, which is how a wrong axis or a missed fixed joint shows up.
"""

import numpy as np
from urdf_parser_py.urdf import URDF


def rpy_to_R(r, p, y):
    cr, sr, cp, sp, cy, sy = (np.cos(r), np.sin(r), np.cos(p),
                              np.sin(p), np.cos(y), np.sin(y))
    return np.array([
        [cy * cp, cy * sp * sr - sy * cr, cy * sp * cr + sy * sr],
        [sy * cp, sy * sp * sr + cy * cr, sy * sp * cr - cy * sr],
        [-sp, cp * sr, cp * cr]])


def T_of(xyz, rpy):
    T = np.eye(4)
    T[:3, :3] = rpy_to_R(*rpy)
    T[:3, 3] = xyz
    return T


def axis_R(axis, q):
    a = np.asarray(axis, dtype=float)
    a = a / np.linalg.norm(a)
    K = np.array([[0, -a[2], a[1]], [a[2], 0, -a[0]], [-a[1], a[0], 0]])
    return np.eye(3) + np.sin(q) * K + (1 - np.cos(q)) * (K @ K)


class Chain:
    """Chain from `base` to `tip`, with the movable joints in order."""

    def __init__(self, urdf_xml, base, tip):
        self.robot = URDF.from_xml_string(urdf_xml)
        self.links = {l.name: l for l in self.robot.links}
        self.parent_joint = {j.child: j for j in self.robot.joints}
        self.steps = []          # (T_origin, axis or None) from base to tip
        self.joint_names = []
        node, chain = tip, []
        while node != base:
            j = self.parent_joint.get(node)
            if j is None:
                raise ValueError("no path from %s up to %s (stuck at %s)"
                                 % (tip, base, node))
            chain.append(j)
            node = j.parent
        for j in reversed(chain):
            org = j.origin
            T = T_of(org.xyz or [0, 0, 0], org.rpy or [0, 0, 0])
            if j.type in ("revolute", "continuous"):
                self.steps.append((T, np.asarray(j.axis, dtype=float)))
                self.joint_names.append(j.name)
            else:
                self.steps.append((T, None))

    def fk_all(self, q):
        """Returns (T_base_tip, [(z_i, p_i)]) with joint axes in base frame."""
        T = np.eye(4)
        axes = []
        k = 0
        for T_org, axis in self.steps:
            T = T @ T_org
            if axis is not None:
                z = T[:3, :3] @ (axis / np.linalg.norm(axis))
                axes.append((z, T[:3, 3].copy()))
                R = np.eye(4)
                R[:3, :3] = axis_R(axis, q[k])
                T = T @ R
                k += 1
        return T, axes

    def fk(self, q):
        return self.fk_all(q)[0]

    def jacobian(self, q):
        """Geometric Jacobian of the tip, expressed in the base frame."""
        T, axes = self.fk_all(q)
        p_tip = T[:3, 3]
        J = np.zeros((6, len(axes)))
        for i, (z, p) in enumerate(axes):
            J[:3, i] = np.cross(z, p_tip - p)
            J[3:, i] = z
        return J

File: src/blindspot_cpp/src/kin.cpp
#include "blindspot/kin.hpp"

#include <cmath>
#include <map>
#include <memory>
#include <stdexcept>

#include <urdf/model.hpp>

namespace blindspot
{

Eigen::Matrix3d rpy_to_R(double r, double p, double y)
{
  const double cr = std::cos(r);
  const double sr = std::sin(r);
  const double cp = std::cos(p);
  const double sp = std::sin(p);
  const double cy = std::cos(y);
  const double sy = std::sin(y);
  Eigen::Matrix3d R;
  R << cy * cp, cy * sp * sr - sy * cr, cy * sp * cr + sy * sr,
    sy * cp, sy * sp * sr + cy * cr, sy * sp * cr - cy * sr,
    -sp, cp * sr, cp * cr;
  return R;
}

Eigen::Matrix4d T_of(const Eigen::Vector3d & xyz, const Eigen::Vector3d & rpy)
{
  Eigen::Matrix4d T = Eigen::Matrix4d::Identity();
  T.topLeftCorner<3, 3>() = rpy_to_R(rpy(0), rpy(1), rpy(2));
  T.topRightCorner<3, 1>() = xyz;
  return T;
}

Eigen::Matrix3d axis_R(const Eigen::Vector3d & axis, double q)
{
  const Eigen::Vector3d a = axis.normalized();
  Eigen::Matrix3d K;
  K << 0.0, -a(2), a(1),
    a(2), 0.0, -a(0),
    -a(1), a(0), 0.0;
  return Eigen::Matrix3d::Identity() + std::sin(q) * K + (1.0 - std::cos(q)) * (K * K);
}

Chain::Chain(const std::string & urdf_xml, const std::string & base, const std::string & tip)
{
  urdf::Model model;
  if (!model.initString(urdf_xml)) {
    throw std::runtime_error("could not parse the URDF");
  }

  std::map<std::string, urdf::JointConstSharedPtr> parent_joint;
  for (const auto & [name, joint] : model.joints_) {
    (void)name;
    parent_joint[joint->child_link_name] = joint;
  }

  // Walk up from the tip, then replay the chain base-first.
  std::vector<urdf::JointConstSharedPtr> chain;
  std::string node = tip;
  while (node != base) {
    const auto it = parent_joint.find(node);
    if (it == parent_joint.end()) {
      throw std::runtime_error(
              "no path from " + tip + " up to " + base + " (stuck at " + node + ")");
    }
    chain.push_back(it->second);
    node = it->second->parent_link_name;
  }

  for (auto it = chain.rbegin(); it != chain.rend(); ++it) {
    const auto & j = *it;
    const auto & origin = j->parent_to_joint_origin_transform;
    double r = 0.0;
    double p = 0.0;
    double y = 0.0;
    // urdfdom stores the origin as a quaternion; rpy is what the URDF wrote,
    // so go back through it rather than composing a second representation.
    origin.rotation.getRPY(r, p, y);
    const Eigen::Vector3d xyz(origin.position.x, origin.position.y, origin.position.z);
    const Eigen::Matrix4d T = T_of(xyz, Eigen::Vector3d(r, p, y));

    Step step;
    step.T_origin = T;
    if (j->type == urdf::Joint::REVOLUTE || j->type == urdf::Joint::CONTINUOUS) {
      step.axis = Eigen::Vector3d(j->axis.x, j->axis.y, j->axis.z);
      step.movable = true;
      joint_names_.push_back(j->name);
    } else {
      step.movable = false;
    }
    steps_.push_back(step);
  }
}

std::pair<Eigen::Matrix4d, std::vector<Chain::Axis>> Chain::fk_all(
  const Eigen::Ref<const Eigen::VectorXd> & q) const
{
  if (q.size() < static_cast<Eigen::Index>(joint_names_.size())) {
    throw std::invalid_argument("fk_all: too few joint positions for this chain");
  }
  Eigen::Matrix4d T = Eigen::Matrix4d::Identity();
  std::vector<Axis> axes;
  Eigen::Index k = 0;
  for (const Step & step : steps_) {
    T = T * step.T_origin;
    if (step.movable) {
      Axis a;
      a.z = T.topLeftCorner<3, 3>() * step.axis.normalized();
      a.p = T.topRightCorner<3, 1>();
      axes.push_back(a);
      Eigen::Matrix4d R = Eigen::Matrix4d::Identity();
      R.topLeftCorner<3, 3>() = axis_R(step.axis, q(k));
      T = T * R;
      ++k;
    }
  }
  return {T, axes};
}

Eigen::Matrix4d Chain::fk(const Eigen::Ref<const Eigen::VectorXd> & q) const
{
  return fk_all(q).first;
}

Eigen::MatrixXd Chain::jacobian(const Eigen::Ref<const Eigen::VectorXd> & q) const
{
  const auto [T, axes] = fk_all(q);
  const Eigen::Vector3d p_tip = T.topRightCorner<3, 1>();
  Eigen::MatrixXd J(6, static_cast<Eigen::Index>(axes.size()));
  for (size_t i = 0; i < axes.size(); ++i) {
    const Eigen::Index c = static_cast<Eigen::Index>(i);
    J.block<3, 1>(0, c) = axes[i].z.cross(p_tip - axes[i].p);
    J.block<3, 1>(3, c) = axes[i].z;
  }
  return J;
}

}  // namespace blindspot
